TLSv1.3: Support KeyUpdate

mattcaswell opened the pull request:
<!--
Thank you for your pull request. Please review below requirements.

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING
-->

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->
- [x] documentation is added or updated
- [ ] tests are added or updated

##### Description of change
<!-- Provide a description of the changes.

If it fixes a github issue, add Fixes #XXXX.
-->
This adds support for the TLSv1.3 KeyUpdate message.

It adds a new API call `SSL_key_update()` which provides the ability for an application to update its sending keys, and send a KeyUpdate message. Additionally the application can request that the peer also updates its own sending keys at the same time (SSL_KEY_UPDATE_REQUESTED) or not (SSL_KEY_UPDATE_NOT_REQUESTED).

At one point it was suggested that we could make a call to `SSL_renegotiate()` do a KeyUpdate instead for TLSv1.3. On reflection I'm not convinced that's a great idea - so I haven't done that. Renegotiation and KeyUpdate are different things. For example, you can't do client auth through KeyUpdate. It would confuse the semantics of the function to try and fit the two concepts into a single API call. It also wouldn't remove the need for a separate `SSL_key_update()` API anyway because of the need to flag whether a peer update is requested or not.

It occurs to me that an application is going to need to know whether a TLSv1.3 (or greater) connection has been established or not (e.g. to know whether to bother calling SSL_renegotiate() or not). So I have also provided a `SSL_is_tls13_plus()` function. Not sure that is the best name. Open to suggestions.

I have managed to (manually) test sending KeyUpdates against Boring but not receiving them, since Boring never sends them at the moment. Picotls also can't handle them. It does work OpenSSL talking to OpenSSL though.

This is marked as WIP as I still need to provide some documentation for the new API call. I also need to write some tests.

richsalz commented:
I totally agree with your intro comment about keyupdate/renegotiation.

richsalz commented on apps/s_client.c:
```diff
@@ -2537,6 +2538,15 @@ int s_client_main(int argc, char **argv)
                 cbuf_len = 0;
             }
+
+            if ((!c_ign_eof) && ((cbuf[0] == 'K' || cbuf[0] == 'k' )
```
Sigh.  I should have fixed up that stuff before (extra parens).  Rats.

mattcaswell replied:
I removed the extra parens for this instance.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -675,6 +679,11 @@ int SSL_is_dtls(const SSL *s)
     return SSL_IS_DTLS(s) ? 1 : 0;
 }
 
+int SSL_is_tls13_plus(const SSL *s)
```
why is this named ```_plus```?

mattcaswell replied:
It was an attempt at future proofing, so that it isn't confusing when we have TLSv1.4. The idea was that it should return 1 if you have a TLSv1.3 or later protocol. I'm not especially happy with the name, except to say that I think `SSL_is_tls13()` is worse.

mattcaswell replied:
This function is only tangentially relevant to this PR so, as it is contentious, I have removed it for now and inserted a TODO asking the question about how applications determine whether they should call `SSL_key_update()` or `SSL_renegotiate()`. To be dealt with in a future PR.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -1714,14 +1723,42 @@ int SSL_shutdown(SSL *s)
+    }
+
+    ossl_statem_set_in_init(s, 1);
+
```
delete this blank line and the one before the return

mattcaswell replied:
Done

levitte commented:
Instead of `SSL_is_tls13_plus()`, I'd suggest checking for capabilities.  `SSL_can_key_update()` or something more elegant.

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -1714,14 +1723,42 @@ int SSL_shutdown(SSL *s)
+    return 1;
+}
+
+SSL_KEY_UPDATE SSL_get_key_update_type(SSL *s)
```
is this really going to be needed?

FdaSilvaYY replied:
`SSL_KEY_UPDATE SSL_get_key_update_type(const SSL *s)` ?
Enum as API return type, are you sure ?

mattcaswell replied:
Why is that a problem? Although, actually, I am pondering removing this function anyway.

mattcaswell replied:
I already used it in s_client (that's why I added it).

richsalz commented on ssl/ssl_locl.h:
```diff
@@ -993,6 +995,10 @@ struct ssl_st {
     EVP_CIPHER_CTX *enc_write_ctx; /* cryptographic state */
     unsigned char write_iv[EVP_MAX_IV_LENGTH]; /* TLSv1.3 static write IV */
     EVP_MD_CTX *write_hash;     /* used for mac generation */
+
```
i dislike blank lines within a struct.

mattcaswell replied:
I prefer them - I think it helps readability. But I don't care enough about it, so I removed them.

richsalz commented on ssl/statem/statem_lib.c:
```diff
@@ -491,6 +491,65 @@ int tls_construct_finished(SSL *s, WPACKET *pkt)
+    }
+
+    s->key_update = SSL_KEY_UPDATE_NONE;
+
```
move this blank line to before the err label.

mattcaswell replied:
Done

richsalz commented on ssl/statem/statem_lib.c:
```diff
@@ -491,6 +491,65 @@ int tls_construct_finished(SSL *s, WPACKET *pkt)
+    return 0;
+}
+
+#define MAX_KEY_UPDATE_MESSAGES     32
```
think this should  be in a header file, not "hidden" in the source.

mattcaswell replied:
I moved it to statem_locl.h

mattcaswell commented:
> Instead of SSL_is_tls13_plus(), I'd suggest checking for capabilities. SSL_can_key_update() or something more elegant.

That's an interesting idea which is quite attractive. The main problem I see with that is that there are a whole load of reasons why you're going to need to know you're dealing with a TLSv1.3 connection and most of them are nothing to do with KeyUpdate. For example the way sessions work is quite different, client auth, early data, etc. Should we have a whole suite of `SSL_can_xxx()` functions which all do the same thing - and tell you whether you have TLSv1.3 or not.

levitte commented:
> > Instead of SSL_is_tls13_plus(), I'd suggest checking for capabilities. SSL_can_key_update() or something more elegant.

> That's an interesting idea which is quite attractive. The main problem I see with that is that there are a whole load of reasons why you're going to need to know you're dealing with a TLSv1.3 connection and most of them are nothing to do with KeyUpdate. For example the way sessions work is quite different, client auth, early data, etc. Should we have a whole suite of `SSL_can_xxx()` functions which all do the same thing - and tell you whether you have TLSv1.3 or not.

I would prefer that.  I'm thinking future.  What if a TLSv1.4 also gets numerous incomatible changes?  TLSv2.0?  That smells version hell galore, and may screw things up quite badly for those who don't remember precisely which version has exactly what capabilities.  It's a problem waiting to happen.  However, asking "can I do this?" and get a precise response for exactly that thing someone wants to do should be rather safe.

richsalz commented:
Please don't add these functions until (unless) we need them.

levitte commented:
Well, @mattcaswell has identified a need for one now, so...

richsalz commented:
No he hasn't.  He thinks it's useful for future-proofing.  My counter to that is "why isn't getting the TLS version enough?"  We should not add API's we think might be useful, as we have to live with them for a very very long time.

mattcaswell commented:
> "why isn't getting the TLS version enough?"

Well I did think about that. The "obvious" way to do it is:

````
if (SSL_get_version(s) == TLS1_3_VERSION)
    do_something();
````
Of course that fails as soon as TLSv1.4 comes along. So perhaps better is:
````
if (SSL_get_version() >= TLS1_3_VERSION)
    do_something();
````
But actually that's worse because the above evaluates to "true" if you happen to be using DTLS (any version). So actually this is quite tricky to get right and probably the kind of thing that *should* be the responsibility of the library.

richsalz commented:
Oh come on :)  The only test that matters right now is "is this 1.3"
NOBODY needs to ask "TLS 1.3 or later"  No bo dy.

kaduk commented:
Why should the application need to check whether it can do something before trying to do it?  Let the application just SSL_keyupdate() and deal with the error if the connection is TLS 1.2 or below.

mattcaswell commented:
@richsalz wrote:
> Oh come on :) The only test that matters right now is "is this 1.3"
NOBODY needs to ask "TLS 1.3 or later" No bo dy.

Well, that's true *right now*. But that's the point isn't it? Code written in applications *right now* is potentially going to live for a long time....past the point when TLSv1.4 comes out. And actually we are talking about an API with the lifespan of 1.1.1, so whatever we decide now has to still make sense for applications written in 2+ years time.

Even if you don't buy the argument above and *we* don't think people need to worry about TLSv1.4 yet, you can guarantee that people are going to write:
`if (SSL_get_version() >= TLS1_3_VERSION)`

Which is wrong.

Actually though Richard is right. *Right now* applications *do* need to know whether they should call SSL_key_update() or SSL_renegotiate()...and perhaps asking if the version is TLSv1.3 or greater is the wrong test anyway (what about when the extension comes along that disables KeyUpdate requests; or when TLSv1.4 removes the KeyUpdate capability). So I'm quite drawn to the capability test idea - although I don't like the potential explosion of different functions that would entail (perhaps a single function that takes as a parameter a feature id?)

@kaduk wrote:
> Why should the application need to check whether it can do something before trying to do it? Let the application just SSL_keyupdate() and deal with the error if the connection is TLS 1.2 or below.

Yes that is possible too, and also something I thought about. Its quite messy though...typically this is going to end up putting error messages on the queue which you probably want to remove before continuing. Also you can't tell whether it failed because the connection doesn't support it or because "something bad happened".

richsalz commented:
By adding this API, you are encouraging developers to predict the future, which is a very bad idea.

mattcaswell commented:
> By adding this API, you are encouraging developers to predict the future, which is a very bad idea.

Which one? Do you mean `SSL_is_tls13_plus()` or Richard's capability test idea? If the latter then I disagree...deciding whether to call `SSL_key_update()` or `SSL_renegotiate()` is a "right now" problem that has nothing to do with what might happen in the future (although it could be impacted by those things).

richsalz commented:
The only test that is needed right now is "is this TLS 1.3"  
We don't know what TLS-next will even be called.
We don't know what TLS-next will decide; it could do PFS and new keys for every record.
What do app developers need *right now*?

mattcaswell commented:
> What do app developers need right now?

They need to know whether to call SSL_key_update() or SSL_renegotiate() (and probably various other similar things).

richsalz commented:
Skipping the "and probably" since that's a guess, not a fact :)

A single API that says "SSL_refresh_keys" which calls the right thing underneath seems like a reasonable approach, then.

mattcaswell commented:
> Skipping the "and probably" since that's a guess, not a fact :)

Well one other thing that immediately springs to mind is session handling. With <= TLSv1.2 you can call SSL_get0_session() at any time after the connection is established. With TLSv1.3 if you do that before the NewSessionTicket message has been sent then you are going to get a bad session (this is an outstanding TODO item). Another one will be around whether you can do early data or not.

> A single API that says "SSL_refresh_keys" which calls the right thing underneath seems like a reasonable approach, then.

Hmmm....that's a possibility - although it kind of takes us back to the "single function that does two different things" argument that made me want to separate SSL_key_update() from SSL_renegotiate() in the first place.

richsalz commented:
It's not quite the same, it moves up a level and only does the same thing on both versions.  I deliberately called it "refresh keys"

Early data is a tls1.3-or-not thing.  not sure if session keys matter, have to see what happens when you finish that code.

levitte commented:
@mattcaswell said:
> So I'm quite drawn to the capability test idea - although I don't like the potential explosion of different functions that would entail (perhaps a single function that takes as a parameter a feature id?)

Yes, I was thinking that would be a possibility.  It's a bit like the cpuID bits (although we'd keep that internal).

I don't quite understand how specifying *current* capabilities is trying to predict the future.  Do you mean it's an attempt to predict a future where there will be other capabilities?  Considering the history of SSL/TLS so far, is that even a difficult prediction?

However, if eveyone is fine with something like `SSH_refresh_keys()`, I have no issues

mattcaswell commented:
> However, if eveyone is fine with something like SSH_refresh_keys(), I have no issues

`SSH_refresh_keys()`? Did we change projects :-)

Actually the more I think about it the less convinced I am that `SSL_refresh_keys()` is the right answer.

mattcaswell commented:
Updated to address the comments received so far. Also rebased to deal with some conflicts that were showing. I have also added documentation to this PR. Still needs tests, so still WIP.

mattcaswell commented:
Tests now added. I also rebased. I've taken this out of WIP - this is now ready for review.

richsalz commented on doc/man3/SSL_key_update.pod:
```diff
@@ -0,0 +1,119 @@
+
+ /* TLSv1.3 KeyUpdate message types */
+ typedef enum {
+     /* -1 used so that this is an invalid value for the on-the-wire protocol */
```
I think we decided that public API's must return int because the sizeof an enum could change if new values are added.  I tried to argue against this, and added a check for it in sanitytest, but lost anyway.

mattcaswell replied:
Hmmm. If we're going to say that, then probably we should not also use the enum for the input parameter to SSL_key_update(). They are the same thing so they should be the same type - if one is an enum then so is the other. If one is an int then they are both ints. That doesn't stop us from treating it as an enum internally of course.

richsalz replied:
yes, an int in both parts.  sad.

mattcaswell replied:
OTOH we would have to add a *lot* of KeyUpdate types for the sizeof to change...like we have 3 at the moment :-)

mattcaswell replied:
Ok. int it is.

richsalz replied:
it's not just us; a compiler could change it across releases...

richsalz commented on doc/man3/SSL_key_update.pod:
```diff
@@ -0,0 +1,119 @@
+=head1 DESCRIPTION
+
+SSL_key_update() schedules an update of the keys for the current TLS connection.
+If the B<updatetype> parameter is set to B<SSL_KEY_UPDATE_NOT_REQUESTED> then
```
Maybe ```SSL_KEY_UPDATE_PEER_OK``` and ```SSL_KEY_UDPATE_PEER_REQUESTED``` are better names?

mattcaswell replied:
I guess I should make these "#define" values now, given that we're not using an enum anymore. The UPDATE_REQUESTED and UPDATE_NOT_REQUESTED names are what they are called in the RFC so I thought we should maintain consistency.

richsalz replied:
that's an excellent point, keep it!

richsalz commented on ssl/ssl_lib.c:
```diff
@@ -1714,14 +1718,45 @@ int SSL_shutdown(SSL *s)
     /*
-     * TODO(TLS1.3): Return an error for now. Perhaps we should do a KeyUpdate
-     * instead when we support that?
+     * TODO(TLS1.3): How will applications know whether TLSv1.3+ has been
```
SSL_get_version?  And remove the "+" from 1.3

mattcaswell replied:
Well, that's the discussion we had before!! My compromise was to add this TODO instead of the SSL_is_tls13_plus() function (deferring that to another PR as its more of a tangential issue). And of course we disagree on whether we should be thinking TLSv1.3 or TLSv1.3+....but as its only a TODO comment I guess I can compromise on the compromise. I'll still be thinking TLSv1.3+ when I come to look at this again though ;-)

richsalz commented on ssl/statem/statem_clnt.c:
```diff
@@ -719,6 +731,14 @@ WORK_STATE ossl_statem_client_post_work(SSL *s, WORK_STATE wst)
+    case TLS_ST_CW_KEY_UPDATE:
+        if (statem_flush(s) != 1)
+            return WORK_MORE_A;
+
```
prefer you delete this blank line.

richsalz commented on ssl/statem/statem_srvr.c:
```diff
@@ -822,6 +840,14 @@ WORK_STATE ossl_statem_server_post_work(SSL *s, WORK_STATE wst)
+    case TLS_ST_SW_KEY_UPDATE:
+        if (statem_flush(s) != 1)
+            return WORK_MORE_A;
+
```
and this one.

mattcaswell commented:
New commits pushed addressing all comments.

richsalz commented:
+1

mattcaswell commented:
Pushed. Thanks.
